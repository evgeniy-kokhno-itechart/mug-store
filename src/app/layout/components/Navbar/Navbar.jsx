import React from 'react';
import { NavLink } from 'react-router-dom';
import { PropTypes } from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { CurrenciesDropdown } from '../../../currency';
import { CartService } from '../../../cart';
import { LoginLogoutBar } from '../../../shared';
import './Navbar.scss';

const Navbar = ({
  currentUserName,
  currencyState,
  cart,
  changeCurrency,
  logoutUser,
}) => {
  const cartCount = CartService.countProducts(cart);

  return (
    <nav className="nav-bar">
      <input className="nav-bar__checkbox" id="nav-toggle" type="checkbox" />
      <label htmlFor="nav-toggle" className="nav-bar__burger-btn" role="button">
        <FontAwesomeIcon icon="bars" />
      </label>

      <div className="nav-bar__collapse-wrapper">
        <div className="nav-bar__left">
          <NavLink className="navbrand" to="/">
            Mug Store
          </NavLink>

          <NavLink to="/catalog" className="navlink" role="navigation">
            Catalog
          </NavLink>

          <NavLink to="/about" className="navlink" role="navigation">
            About
          </NavLink>

          <NavLink to="/cart" className="navlink" role="navigation">
            Cart
            {cartCount !== 0
              ? <span className="cart-count-badge">{cartCount}</span>
              : null}
          </NavLink>
        </div>

        <div className="nav-bar__right">
          <CurrenciesDropdown currencyState={currencyState} onChange={changeCurrency} />
          <LoginLogoutBar currentUserName={currentUserName} logoutUser={logoutUser} />
        </div>
      </div>
    </nav>
  );
};

Navbar.propTypes = {
  cart: PropTypes.arrayOf(PropTypes.shape({ quantity: PropTypes.number })),
  currentUserName: PropTypes.string,

  currencyState: PropTypes.shape({
    currencies: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })),
    currenciesStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
    currentCurrency: PropTypes.shape({ id: PropTypes.string }),
  }),
  changeCurrency: PropTypes.func.isRequired,

  logoutUser: PropTypes.func.isRequired,
};

Navbar.defaultProps = {
  currencyState: {
    currencies: [],
    currenciesStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
    currentCurrency: { id: '0' },
  },
  currentUserName: '',
  cart: [],
};

export default Navbar;
