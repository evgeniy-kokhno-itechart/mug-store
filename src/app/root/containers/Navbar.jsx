import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { changeCurrency, CurrenciesDropdown } from '../../currency';
import LoginLogoutPanel from './LoginLogoutBar';

export const Navbar = ({
  currentCurrency,
  cart,
  isCurrenciesLoading,
  hasCurrenciesLoadFailed,
  errorCurrenciesLoading,
  currencies: currencyOptions,
  // eslint-disable-next-line no-shadow
  changeCurrency,
}) => {
  const cartCount = cart.reduce((sumQty, currentProduct) => sumQty + currentProduct.quantity, 0);

  return (
    <nav className="navbar navbar-expand-md navbar-dark bg-dark sticky-top">
      <button
        className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#mugStoreNavbar"
        aria-controls="mugStoreNavbar"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span className="navbar-toggler-icon" />
      </button>

      <div className="collapse navbar-collapse" id="mugStoreNavbar">
        <NavLink className="navbar-brand" to="/">
          Mug Store
        </NavLink>

        <div className="navbar-nav">
          <NavLink to="/catalog" className=" nav-item nav-link">
            Catalog
          </NavLink>

          <NavLink to="/about" className="nav-item nav-link">
            About
          </NavLink>

          <NavLink to="/cart" className="nav-item nav-link">
            Cart
            {cartCount !== 0 ? <span className="badge badge-warning ml-1">{cartCount}</span> : null}
          </NavLink>
        </div>
      </div>

      <div>
        <CurrenciesDropdown
          currentCurrencyId={currentCurrency.id}
          currencyOptions={currencyOptions}
          onCurrencyChange={changeCurrency}
          isCurrenciesLoading={isCurrenciesLoading}
          hasLoadFailed={hasCurrenciesLoadFailed}
          errorMessage={errorCurrenciesLoading}
        />

        <LoginLogoutPanel />
      </div>
    </nav>
  );
};

Navbar.propTypes = {
  currentUser: PropTypes.shape({ name: PropTypes.string }),
  currentCurrency: PropTypes.shape({ id: PropTypes.string }).isRequired,
  isCurrenciesLoading: PropTypes.bool.isRequired,
  hasCurrenciesLoadFailed: PropTypes.bool,
  errorCurrenciesLoading: PropTypes.string,
  cart: PropTypes.arrayOf(PropTypes.shape({ quantity: PropTypes.number })),
  currencies: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })),
  changeCurrency: PropTypes.func.isRequired,
};

Navbar.defaultProps = {
  hasCurrenciesLoadFailed: false,
  errorCurrenciesLoading: '',
  currencies: [],
  currentUser: { roles: [] },
  cart: [],
};

const mapStateToProps = state => ({
  cart: state.cart.cart,
  currentUser: state.user.currentUser,
  currencies: state.currency.currencies,
  isCurrenciesLoading: state.currency.currenciesStatus.isGettingCurrenciesInProcess,
  hasCurrenciesLoadFailed: state.currency.currenciesStatus.hasGettingCurrenciesFailed,
  errorCurrenciesLoading: state.currency.currenciesStatus.error,
  currentCurrency: state.currency.currentCurrency,
});

const mapDispatchToProps = {
  changeCurrency,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Navbar);
