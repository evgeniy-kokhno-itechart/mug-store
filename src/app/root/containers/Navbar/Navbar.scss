@import "Variables";
@import "Mixins";

.nav-bar {
    position: sticky;
    top: 0;
    padding: 0.5rem 1rem;
    z-index: 1000;
    background-color: $color-primary-dark;

    @include respond(phone) {
        display: flex;
        align-items: center;

        &__checkbox:checked ~ &__collapse-wrapper {
            height: 100%;
            visibility: visible;
            opacity: 1;
        }
    }

    &__checkbox {
        display: none;
    }

    &__collapse-wrapper {
        display: flex;
        justify-content: space-between;
        align-items: center;
        transition: all 0.5s;
        
        @include respond(phone) {
            margin: 0 1rem;
            justify-content: space-around;
            flex-wrap: wrap;
            visibility: hidden;
            height: 0;
            opacity: 0;
        }
    }

    &__burger-btn {
        font-size: 2rem;
        text-align: center;
        color: $color-white;
        background-image: linear-gradient(to right bottom, rgba($color-gray-dark, 0.7), rgba($color-gray-dark, 1));
        border: 1px solid black;
        min-width: 3rem;
        display: none;

        @include respond(phone) {
            display: inline-block;
            margin: 0;
        }
    }

    &__left {
        @include respond(phone) {
            display: flex;
            align-items: center;
        }
    }
}

.navbrand {
    font-size: 1.1rem;
    color: $color-white;
    margin: 0 0.7rem 0 0;

    &:hover {
        color: $color-white;
        text-decoration: none;
    }

    @include respond(phone) {
        font-size: inherit;
        margin-left: 0;
    }
}

.navlink {
    color: $color-gray-light;
    margin: 0.5rem;

    &:hover {
        text-decoration: none;
        color: $color-white;
    }

    &.active {
        color: $color-white;
    }
}

.cart-count-badge {
    background-color: $color-yellow;
    color: $color-primary-dark;
    padding: 0.1rem 0.3rem;
    margin: 0 0.3rem;
    border-radius: 3px;
}