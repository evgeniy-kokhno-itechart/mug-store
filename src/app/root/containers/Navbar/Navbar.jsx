import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { currencyActions, CurrenciesDropdown } from '../../../currency';
import { userActions } from '../../../user';
import { CartService } from '../../../cart';
import LoginLogoutBar from '../../components/LoginLogoutBar';
import './Navbar.scss';

export const Navbar = ({
  currentUserName,
  currencyState,
  cart,
  changeCurrency,
  logoutUser,
}) => {
  const cartCount = CartService.countProducts(cart);

  return (
    <nav className="nav-bar">
      <input className="nav-bar__checkbox" id="nav-toggle" type="checkbox" />
      <label htmlFor="nav-toggle" className="nav-bar__burger-btn" role="button">
        <FontAwesomeIcon icon="bars" />
      </label>

      <div className="nav-bar__collapse-wrapper">
        <div className="nav-bar__left">
          <NavLink className="navbrand" to="/">
            Mug Store
          </NavLink>

          <NavLink to="/catalog" className="navlink" role="navigation">
            Catalog
          </NavLink>

          <NavLink to="/about" className="navlink" role="navigation">
            About
          </NavLink>

          <NavLink to="/cart" className="navlink" role="navigation">
            Cart
            {cartCount !== 0
              ? <span className="cart-count-badge">{cartCount}</span>
              : null}
          </NavLink>
        </div>

        <div className="nav-bar__right">
          <CurrenciesDropdown currencyState={currencyState} onChange={changeCurrency} />
          <LoginLogoutBar currentUserName={currentUserName} logoutUser={logoutUser} />
        </div>
      </div>
    </nav>
  );
};

Navbar.propTypes = {
  cart: PropTypes.arrayOf(PropTypes.shape({ quantity: PropTypes.number })),
  currentUserName: PropTypes.string,

  currencyState: PropTypes.shape({
    currencies: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })),
    currenciesStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
    currentCurrency: PropTypes.shape({ id: PropTypes.string }),
  }),
  changeCurrency: PropTypes.func.isRequired,

  logoutUser: PropTypes.func.isRequired,
};

Navbar.defaultProps = {
  currencyState: {
    currencies: [],
    currenciesStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
    currentCurrency: { id: '0' },
  },
  currentUserName: '',
  cart: [],
};

const mapStateToProps = state => ({
  cart: state.cart.cart,
  currencyState: state.currency,
  currentUserName: state.user.currentUser.name,
});

const mapDispatchToProps = {
  changeCurrency: currencyActions.ChangeCurrency,
  logoutUser: userActions.Logout.InitiateApiCall,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Navbar);
