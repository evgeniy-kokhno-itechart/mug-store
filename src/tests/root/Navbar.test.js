/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Navbar } from '../../app/root/containers/Navbar/Navbar';

configure({ adapter: new Adapter() });

describe('<Navbar />', () => {
  let NavbarWrapper;

  beforeEach(() => {
    const changeCurrency = jest.fn();
    const logoutUser = jest.fn();
    NavbarWrapper = shallow(
      <Navbar
        currentCurrency={{ id: '0' }}
        isCurrenciesLoading
        changeCurrency={changeCurrency}
        logoutUser={logoutUser}
      />,
    );
  });

  it('renders properly with default props', () => {
    expect(NavbarWrapper).toMatchSnapshot();
  });

  it('renders user name with currentUser provided', () => {
    NavbarWrapper.setProps({ currentUserName: 'testUsername' });
    expect(NavbarWrapper).toMatchSnapshot();
  });

  it('renders with currenciesStatus.hasFailed and error provided', () => {
    NavbarWrapper.setProps({
      currencyState: {
        currenciesStatus: { isInProcess: false, hasFailed: true, error: 'test error' },
      },
    });
    expect(NavbarWrapper).toMatchSnapshot();
  });

  it('renders cart count badge with cart provided', () => {
    NavbarWrapper.setProps({ cart: [{ quantity: 3 }, { quantity: 1 }, { quantity: 7 }] });
    expect(NavbarWrapper).toMatchSnapshot();
  });

  it('renders currencies dropdown filled with currencies list when provided', () => {
    NavbarWrapper.setProps({
      currencyState: {
        currencies: [{ id: '1', name: 'CURR1' }, { id: '2', name: 'CURR2' }],
        currenciesStatus: { isInProcess: false, hasFailed: false, error: '' },
        currentCurrency: { id: '2' },
      },
    });
    expect(NavbarWrapper).toMatchSnapshot();
  });
});
