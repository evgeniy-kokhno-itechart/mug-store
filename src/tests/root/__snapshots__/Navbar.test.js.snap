// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> renders cart count badge with cart provided 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark sticky-top"
>
  <button
    aria-controls="mugStoreNavbar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#mugStoreNavbar"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="mugStoreNavbar"
  >
    <NavLink
      className="navbar-brand"
      to="/"
    >
      Mug Store
    </NavLink>
    <div
      className="navbar-nav"
    >
      <NavLink
        className=" nav-item nav-link"
        to="/catalog"
      >
        Catalog
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/about"
      >
        About
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/cart"
      >
        Cart
        <span
          className="badge badge-warning ml-1"
        >
          11
        </span>
      </NavLink>
    </div>
  </div>
  <div>
    <CurrenciesDropdown
      currencyState={
        Object {
          "currencies": Array [],
          "currenciesStatus": Object {
            "error": "",
            "hasFailed": false,
            "isInProcess": true,
          },
          "currentCurrency": Object {
            "id": "0",
          },
        }
      }
      onChange={[MockFunction]}
    />
    <LoginLogoutBar
      currentUserName=""
      logoutUser={[MockFunction]}
    />
  </div>
</nav>
`;

exports[`<Navbar /> renders currencies dropdown filled with currencies list when provided 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark sticky-top"
>
  <button
    aria-controls="mugStoreNavbar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#mugStoreNavbar"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="mugStoreNavbar"
  >
    <NavLink
      className="navbar-brand"
      to="/"
    >
      Mug Store
    </NavLink>
    <div
      className="navbar-nav"
    >
      <NavLink
        className=" nav-item nav-link"
        to="/catalog"
      >
        Catalog
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/about"
      >
        About
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/cart"
      >
        Cart
      </NavLink>
    </div>
  </div>
  <div>
    <CurrenciesDropdown
      currencyState={
        Object {
          "currencies": Array [
            Object {
              "id": "1",
              "name": "CURR1",
            },
            Object {
              "id": "2",
              "name": "CURR2",
            },
          ],
          "currenciesStatus": Object {
            "error": "",
            "hasFailed": false,
            "isInProcess": false,
          },
          "currentCurrency": Object {
            "id": "2",
          },
        }
      }
      onChange={[MockFunction]}
    />
    <LoginLogoutBar
      currentUserName=""
      logoutUser={[MockFunction]}
    />
  </div>
</nav>
`;

exports[`<Navbar /> renders properly with default props 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark sticky-top"
>
  <button
    aria-controls="mugStoreNavbar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#mugStoreNavbar"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="mugStoreNavbar"
  >
    <NavLink
      className="navbar-brand"
      to="/"
    >
      Mug Store
    </NavLink>
    <div
      className="navbar-nav"
    >
      <NavLink
        className=" nav-item nav-link"
        to="/catalog"
      >
        Catalog
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/about"
      >
        About
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/cart"
      >
        Cart
      </NavLink>
    </div>
  </div>
  <div>
    <CurrenciesDropdown
      currencyState={
        Object {
          "currencies": Array [],
          "currenciesStatus": Object {
            "error": "",
            "hasFailed": false,
            "isInProcess": true,
          },
          "currentCurrency": Object {
            "id": "0",
          },
        }
      }
      onChange={[MockFunction]}
    />
    <LoginLogoutBar
      currentUserName=""
      logoutUser={[MockFunction]}
    />
  </div>
</nav>
`;

exports[`<Navbar /> renders user name with currentUser provided 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark sticky-top"
>
  <button
    aria-controls="mugStoreNavbar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#mugStoreNavbar"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="mugStoreNavbar"
  >
    <NavLink
      className="navbar-brand"
      to="/"
    >
      Mug Store
    </NavLink>
    <div
      className="navbar-nav"
    >
      <NavLink
        className=" nav-item nav-link"
        to="/catalog"
      >
        Catalog
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/about"
      >
        About
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/cart"
      >
        Cart
      </NavLink>
    </div>
  </div>
  <div>
    <CurrenciesDropdown
      currencyState={
        Object {
          "currencies": Array [],
          "currenciesStatus": Object {
            "error": "",
            "hasFailed": false,
            "isInProcess": true,
          },
          "currentCurrency": Object {
            "id": "0",
          },
        }
      }
      onChange={[MockFunction]}
    />
    <LoginLogoutBar
      currentUserName="testUsername"
      logoutUser={[MockFunction]}
    />
  </div>
</nav>
`;

exports[`<Navbar /> renders with currenciesStatus.hasFailed and error provided 1`] = `
<nav
  className="navbar navbar-expand-md navbar-dark bg-dark sticky-top"
>
  <button
    aria-controls="mugStoreNavbar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#mugStoreNavbar"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    className="collapse navbar-collapse"
    id="mugStoreNavbar"
  >
    <NavLink
      className="navbar-brand"
      to="/"
    >
      Mug Store
    </NavLink>
    <div
      className="navbar-nav"
    >
      <NavLink
        className=" nav-item nav-link"
        to="/catalog"
      >
        Catalog
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/about"
      >
        About
      </NavLink>
      <NavLink
        className="nav-item nav-link"
        to="/cart"
      >
        Cart
      </NavLink>
    </div>
  </div>
  <div>
    <CurrenciesDropdown
      currencyState={
        Object {
          "currenciesStatus": Object {
            "error": "test error",
            "hasFailed": true,
            "isInProcess": false,
          },
        }
      }
      onChange={[MockFunction]}
    />
    <LoginLogoutBar
      currentUserName=""
      logoutUser={[MockFunction]}
    />
  </div>
</nav>
`;
